modules/renter/hostdb/negotiate.go

```diff
@@ -443,36 +443,75 @@ func (hdb *HostDB) newHostUploader(settings modules.HostSettings) (*hostUploader
 	return hu, nil
 }
 
+// A HostPool is a collection of hosts used to upload a file.
+type HostPool interface {
+	// UniqueHosts will return up to 'n' unique hosts that are not in 'old'.
+	UniqueHosts(n int, old []modules.NetAddress) []Uploader
+
+	// Close terminates all connections in the host pool.
+	Close() error
+}
+
+type pool struct {
+	hosts []*hostUploader
+	hdb   *HostDB
+}
+
+func (p *pool) Close() error {
+	for _, h := range p.hosts {
+		h.Close()
+	}
+	return nil
+}
+
 // UniqueHosts will return up to 'n' unique hosts that are not in 'old'. Note
-// that this is a blocking call that performs network I/O.
-func (hdb *HostDB) UniqueHosts(n int, old []Uploader) (hosts []Uploader) {
-	// remove old hosts from tree
-	hdb.mu.Lock()
-	var oldEntries []*hostEntry
-	for _, host := range old {
-		node, exists := hdb.activeHosts[host.Address()]
-		if !exists {
-			continue
+// that this may require negotiating new contracts.
+func (p *pool) UniqueHosts(n int, old []modules.NetAddress) (hosts []Uploader) {
```

Comment:
'old' is a confusing variable name, and the docstring should probably be more verbose, as this function gets exported from the package. Explaining how the unique hosts are selected and why is probably a good idea (how at a high level)